k4FWCore/components/UniqueIDGenSvc.cpp

```diff
@@ -18,27 +18,49 @@
  */
 #include "UniqueIDGenSvc.h"
 
-DECLARE_COMPONENT(UniqueIDGenSvc)
+#include <cstdint>
+#include <string>
 
 UniqueIDGenSvc::UniqueIDGenSvc(const std::string& name, ISvcLocator* svcLoc) : base_class(name, svcLoc) {}
 
-StatusCode UniqueIDGenSvc::initialize() {
-  StatusCode sc = Service::initialize();
-  return sc;
-}
-
-const size_t bits32    = std::numeric_limits<uint32_t>::digits;
-const size_t bits64    = std::numeric_limits<uint64_t>::digits;
-const size_t bitsSizeT = std::numeric_limits<size_t>::digits;
+constexpr size_t bits32    = std::numeric_limits<uint32_t>::digits;
+constexpr size_t bits64    = std::numeric_limits<uint64_t>::digits;
+constexpr size_t bitsSizeT = std::numeric_limits<size_t>::digits;
 
 size_t UniqueIDGenSvc::getUniqueID(uint32_t evt_num, uint32_t run_num, const std::string& name) const {
-  std::bitset<bits64>    seed_bits(this->m_seed);
-  std::bitset<bits32>    event_num_bits(evt_num), run_num_bits(run_num);
-  size_t                 str_hash = std::hash<std::string>{}(name);
-  std::bitset<bitsSizeT> name_bits(str_hash);
+  std::bitset<bits64>    seed_bits      = this->m_seed.value();
+  std::bitset<bits32>    event_num_bits = evt_num, run_num_bits = run_num;
+  size_t                 str_hash  = std::hash<std::string>{}(name);
+  std::bitset<bitsSizeT> name_bits = str_hash;
+
+  std::bitset<bits64 + bits32 + bits32 + bitsSizeT> combined_bits;
 
-  std::bitset<bits64 + bits32 + bits32 + bitsSizeT> combined_bits(seed_bits.to_string() + event_num_bits.to_string() +
-                                                                  run_num_bits.to_string() + name_bits.to_string());
+  for (size_t i = 0; i < bitsSizeT; i++) {
+    combined_bits[i] = name_bits[i];
+  }
+  for (size_t i = 0; i < bits32; i++) {
+    combined_bits[i + bitsSizeT] = run_num_bits[i];
+  }
+  for (size_t i = 0; i < bits32; i++) {
+    combined_bits[i + bits32 + bitsSizeT] = event_num_bits[i];
+  }
+  for (size_t i = 0; i < bits64; i++) {
+    combined_bits[i + bits32 + bits32 + bitsSizeT] = seed_bits[i];
+  }
 
-  return std::hash<std::bitset<bits64 + bits32 + bits32 + bitsSizeT>>{}(combined_bits);
+  auto hash     = std::hash<std::bitset<bits64 + bits32 + bits32 + bitsSizeT>>{}(combined_bits);
+  bool inserted = false;
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    std::tie(std::ignore, inserted) = m_uniqueIDs.insert(hash);
+  }
+  if (!inserted) {
+    warning() << "Event number " << evt_num << ", run number " << run_num << " and algorithm name \"" << name
```

Comment:
Maybe this should be fatal, with an option to turn it into a warning? Nobody reads warnings.

Reply:
Is it possible/likely to enter this condition 'by chance' while everything is actually sane? If not then I would be in favor

Reply:
Possible hash collision, but it shouldn't be likely. And if it happens too often, we have a problem we should know about.

Reply:
I changed the default to throw whenever the same ID is asked, with an option not to if needed. Indeed if this is buried in thousands of lines of logs it's not very useful. This will throw if someone runs the same algorithm multiple times with the same name; for example, if digitizing multiple collections with the same digitizer and running it multiple times.